lib/common/socket/evloop.c.h

```diff
@@ -115,19 +115,97 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
     }
 }
 
-static const char *on_read_core(int fd, h2o_buffer_t **input)
+#define NSEC_PER_SEC 1000000000
+
+static inline int64_t
+timespec_to_nsec(const struct timespec *a)
+{
+       return (int64_t)a->tv_sec * NSEC_PER_SEC + a->tv_nsec;
+}
+
+#if defined(__linux__) && !H2O_USE_LIBUV
+static void handle_timestamp(struct st_h2o_evloop_socket_t *sock, struct msghdr *msg)
+{
+    struct scm_timestamping *ts = NULL;
+    uint64_t packet_ts = 0;
+    h2o_loop_t *loop = sock->loop;
+
+    for (struct cmsghdr *cmsg = CMSG_FIRSTHDR(msg); cmsg;
+            cmsg = CMSG_NXTHDR(msg, cmsg)) {
+
+        /* in the future we may want to harvest IP_RECVERR messages here, which could have interesting data */
+        if (cmsg->cmsg_level != SOL_SOCKET)
+            continue;
+
+        switch (cmsg->cmsg_type) {
+            case SO_TIMESTAMPNS:
+                ts = (struct scm_timestamping *)CMSG_DATA(cmsg);
+                packet_ts = timespec_to_nsec(&ts->ts[0]);
+
+                h2o_sliding_counter_start(&loop->packet_latency_nanosec_counter, packet_ts);
+                break;
+            default:
+                /* Ignore other cmsg options */
+                break;
+        }
+    }
+
+    return;
+}
+#else /* !defined(__linux__) || (defined(__linux__) && H2O_USE_LIBUV) */
+static void handle_timestamp(struct st_h2o_evloop_socket_t *sock, struct msghdr *msg)
+{
+    return;
+}
+#endif /* defined(__linux__) && !defined(H2O_USE_LIBUV) */
+
+static const char *on_read_core(struct st_h2o_evloop_socket_t *sock)
 {
     ssize_t read_so_far = 0;
+    int fd = sock->fd;
+    h2o_buffer_t **input = sock->super.ssl == NULL ? &sock->super.input : &sock->super.ssl->input.encrypted;
+    int timestamping = sock->super.timestamping;
+
+    struct msghdr msg;
 
     while (1) {
         ssize_t rret;
         h2o_iovec_t buf = h2o_buffer_try_reserve(input, 4096);
+        size_t buflen = (buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 +1);
+
```

Comment:
How about moving this below, like:
```c
struct iovec iov = { .iov_base = buf.base, .iov_len = buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 + 1 };`
```
The rationale is that we might want to keep memory allocation procedure (h2o_buffer_try_reserve) and failure handling close together

Reply:
yes, i agree the failure handling and the allocation should be closer together. i fixed that.

i left buflen as a separate variable because i personally dislike extremely long / complex expressions (especially with ternaries) because i think they are hard to read... but ultimately up to you 😄  happy to change it to match your desired style.